### ISSUE TITLE
Hyperiond leaks file handles

### ISSUE THREAD

**Issue by penfold42** [hyperion-project/hyperion#688]
If you repeatedly relaunch effects, you'll see lots of .py files left open

How hard would it be to completely re-init the Python engine every time an effect is started ?

This would probably make my clock effect Python 3 PIL library crash problem go away.

It would be even better if we could isolate a Python crash from taking out hyperiond too

**Comment by redPanther** [comment]
Good question. I will investigate ...

**Comment by redPanther** [comment]
please check if this helps:

effect.cpp line 117

fclose(file);

for me it worked ....

**Comment by redPanther** [comment]
@penfold42 can you verify if file handles where closed with with my suggested modifikation.

**Comment by penfold42** [comment]
Will try It over the weekend

**Comment by redPanther** [comment]
Isolating is perhaps not possible, because py script is compiled to binary object during load. The binary code is not separated from hyperion code.
One solution can be that the effects engine is started as a process (fork) instead of thread. Ipc to hyperion can be made via proto.
Then hyperion core won't crash if effect crash and we can separate the code to be more flexible.

**Comment by penfold42** [comment]
I really like this idea - it also stops rogue Python scrips dragging hyperiond to its knees.

How hard would it be to implement ?

I think it will become more important as people bios fancier effect scrips

**Comment by redPanther** [comment]
I think it is manageable. Following steps could be a solution:

First clone src/hyperion-x11 to hyperion-effect and remove x11 grabber code. Then extract effect code and integrate it in hyperion-effect. Add command line options to hyperion-effect.
In hyperiond in the place where effects are started make a fork logic like in hyperiond.cpp for the forwarder thing. Then call new hyperion-effect with effect name and effect path as argument. dont forget to save the  child pid.
If effect should stop, kill the child pid and effect stops.

hyperion-effect should send its output via proto server, like a grabber does.

Not a 5 minute task, but it's managebale. When we have that, then we can remove the grabber code from hyperiond in the same way. Then we have a slim hyperion-core :-) 

For the roadmap: perhaps in hyperion 3? hyperion 2 should move all stuff to qt5

**Comment by brindosch** [comment]
Are we talking about web browser versions or is there a feature level in between?
I didn´t thought to release v2 earlier than 2 or 3 months in future (or more - no deadline).
So just add it to v2 too? v2 breaks anything with some things, no reason to move that fast to release a v3 in 6 months?

**Comment by brindosch** [comment]
After the last release now i cleanup the repo history and than we proceed with v2

**Comment by penfold42** [comment]
We have a lot of numbers between 1.03 and 2.00 !

I'd rather change lots of things like that  break backward compatibility (where we have to) at the same time  in a single major version jump.
Old transforms, qt4, old UDP driver, old effect engine, old git  history old grabber architecture and anything we want to deprecate now should all be done in v2 if we can

With the proposed separate process,is ther a way to gracefully terminate the effect ? Would we simply signal the child process nicely as a replacement for Hyperion.abort ?

**Comment by brindosch** [comment]
To move faster forward we close V1 now, i test currently the last things and push afterwards a new release.
as you listed above there is enough we need to clean up and change. To get a better base (qt5,...) we move immediately to up2date versions of most libs.

I don´t want that the devs (you) waste time to just keep something outdated running (and "hack down" features just to get it)

**Comment by redPanther** [comment]
Hi @penfold42, did you try the patch regarding open file handlers? Would be nice if we can close this issue, because a daemon eating handles isn't funny

**Comment by brindosch** [comment]
@penfold42 ?

**Comment by penfold42** [comment]
I tested it with hyperion-ng and it seems to be fine!

I used hyperion-remote to repeatedly change the effects:
while true; do hyperion-remote -a loungepi:19444 -e "Knight rider" ; hyperion-remote -a loungepi:19444 -e "Cold mood blobs"; done

While its running i changed the file handles with ls -l /proc/32573/fd
There were never more 1 .py file opened.

I did notice memory usage slowly increasing, (as reported by top). I dont know if its a memory leak per se or simply because the new request is needs some malloc before the old ones are freed
